Plugins/NativeCamera/Source/Apple/NativeCameraImpl.mm

```diff
@@ -58,69 +59,138 @@ - (id)init:(std::shared_ptr<Babylon::Plugins::Camera::Impl::ImplData>)implData;
+            {
+                for (AVCaptureDeviceFormat* format in device.formats)
+                {
+                    CMVideoFormatDescriptionRef videoFormatRef = static_cast<CMVideoFormatDescriptionRef>(format.formatDescription);
+                    CMVideoDimensions dimensions = CMVideoFormatDescriptionGetDimensions(videoFormatRef);
+                    
+                    // Reject any resolution that doesn't qualify for the constraint.
+                    if (static_cast<uint32_t>(dimensions.width) > maxWidth || static_cast<uint32_t>(dimensions.height) > maxHeight)
+                    {
+                        continue;
+                    }
+                    
+                    // Calculate pixel count and dimension differential and take the best qualifying one.
+                    uint32_t pixelCount = dimensions.width * dimensions.height;
+                    uint32_t dimDiff = (maxWidth - dimensions.width) + (maxHeight - dimensions.height);
+                    if (bestDevice == NULL || pixelCount > bestPixelCount || (pixelCount == bestPixelCount && dimDiff < bestDimDiff))
+                    {
+                        bestPixelCount = pixelCount;
+                        bestDevice = device;
+                        bestFormat = format;
+                        bestDimDiff = dimDiff;
+                        
+                        // Check if we got an exact match, and exit the loop early in this case.
+                        if (static_cast<uint32_t>(dimensions.width) == maxWidth && static_cast<uint32_t>(dimensions.height) == maxHeight)
+                        {
+                            foundExactMatch = true;
+                            break;
+                        }
+                    }
+                }
                 
-                // In the event that the rear wide angle camera isn't available, default to the front wide angle camera.
-                if (!captureDevice) {
-                    captureDevice = [AVCaptureDevice defaultDeviceWithDeviceType:AVCaptureDeviceTypeBuiltInWideAngleCamera mediaType:AVMediaTypeVideo position:AVCaptureDevicePositionFront];
+                if (foundExactMatch)
+                {
+                    break;
                 }
             }
-#else
-            UNUSED(frontCamera);
-            AVCaptureDevice* captureDevice = [AVCaptureDevice defaultDeviceWithMediaType:AVMediaTypeVideo];
-#endif
-            // Set video capture input: If there a problem initialising the camera, it will give am error.
-            AVCaptureDeviceInput *input = [AVCaptureDeviceInput deviceInputWithDevice:captureDevice error:&error];
+            
+            // If no matching device, throw an error with the message "ConstraintError" which matches the behavior in the browser.
+            if (bestDevice == NULL)
+            {
+                taskCompletionSource.complete(arcana::make_unexpected(std::make_exception_ptr(std::runtime_error{"ConstraintError"})));
```

Comment:
Maybe spell this out more? Instead of "ConstraintError" something like "Could not find a device configuration matching the requested constraings." Or just copy what WebAPI does?

Reply:
This is what the WebAPI does :'(, but I can give it a better error message..

Reply:
Oh well, if it matches WebAPI then I guess that's something!